src/lib/Syntax.hs

```diff
@@ -146,119 +121,123 @@ scalarTableBaseType t = case t of
   BaseTy b  -> b
   _         -> error $ "Not a scalar table: " ++ show t
 
-noEffect :: Effect
-noEffect = Effect mempty Nothing
-
-isPure :: Effect -> Bool
-isPure (Effect eff Nothing) | eff == mempty = True
-isPure _ = False
-
-type LTVar = VarP (LorT Type Kind)
-type ModuleType = ([LTVar], [LTVar])
-data ModuleP body = Module (Maybe BlockId) ModuleType body  deriving (Show, Eq)
-
 -- === front-end language AST ===
 
-data FExpr = FDecl FDecl FExpr
-           | FVar Var
-           | FPrimExpr (PrimExpr Type FExpr FLamExpr)
-           | Annot FExpr Type
-           | SrcAnnot FExpr SrcPos -- TODO: make mandatory?
-             deriving (Eq, Show, Generic)
-
-type Pat = RecTree Var
-data FLamExpr = FLamExpr Pat FExpr  deriving (Show, Eq, Generic)
-type SrcPos = (Int, Int)
-
--- TODO: Unify LetMono and LetPoly. The only real difference is having a type
---       annotation attached and we can handle this with bidirectional try inference.
-data FDecl = LetMono Pat FExpr
-           | LetPoly Var FTLam
-           | TyDef TVar Type
-             deriving (Show, Eq, Generic)
-
-type Var  = VarP Type
-data FTLam = FTLam [TVar] [TyQual] FExpr  deriving (Show, Eq, Generic)
-type FModule = ModuleP [FDecl]
-
-data RuleAnn = LinearizationDef Name    deriving (Show, Eq, Generic)
+type UExpr = WithSrc UExpr'
+data UExpr' = UVar UVar
+            | ULam UBinder UArrow UExpr
+            | UPi  UPiBinder Arrow UType
+            | UApp UArrow UExpr UExpr
+            | UDecl UDecl UExpr
+            | UFor Direction UBinder UExpr
+            | UHole
+            | UTabCon [UExpr] (Maybe UExpr)
+            | UIndexRange (Limit UExpr) (Limit UExpr)
+            | UPrimExpr (PrimExpr Name)
+              deriving (Show, Eq, Generic)
 
--- === normalized core IR ===
+data UDecl = ULet UBinder UExpr  deriving (Show, Eq, Generic)
 
-data Expr = Decl Decl Expr
-          | CExpr CExpr
-          | Atom Atom
-            deriving (Show, Eq, Generic)
+type UType  = UExpr
+type UArrow = ArrowP ()
+type UVar   = VarP ()
 
-data Decl = Let Var CExpr  deriving (Show, Eq, Generic)
+type UPat    = PatP  UVar
+type UPat'   = PatP' UVar
+type UBinder   = (UPat, Maybe UType)
+type UPiBinder = VarP UType
 
-type CExpr = PrimOp  Type Atom LamExpr
-type Con   = PrimCon Type Atom LamExpr
+data UModule = UModule [Name] [Name] [UDecl]  deriving (Show, Eq)
+type SrcPos = (Int, Int)
 
-data Atom = Var Var
-          | TLam [TVar] [TyQual] Expr
-          | Con Con
-            deriving (Show, Eq, Generic)
+type PatP  a = WithSrc (PatP' a)
+data PatP' a = PatBind a
+             | PatPair (PatP a) (PatP a)
+             | PatUnit  deriving (Show, Eq, Functor, Foldable, Traversable)
```

Comment:
Why is this necessary? Patterns should always be non-empty, so I don't think you actually need this?

Reply:
Unit patterns can by useful. `\(). <expr>` can be clearer than `\_. <expr>`.